ui/ts/util/property.ts

```diff
@@ -0,0 +1,119 @@
+// source: util/property.ts
+// Author: Matt Tracy (matt@cockroachlabs.com)
+
+module Utils {
+  "use strict";
+
+  /**
+   * ReadOnlyProperty is a function that returns a value. 
+   *
+   * Each property also has an Epoch() function which returns a monotonically
+   * increasing number indicating the version of the property; if the value
+   * returned by the property changes, the epoch value will be increased.
+   */
+  export interface ReadOnlyProperty<T> {
+    (): T;
+    // Epoch returns the current update epoch for the property. This is a
+    // numeric value that monotonically increases each time the property is
+    // updated.
+    Epoch(): number;
+  }
+
+  /**
+   * A Property is a function which returns a value, but can also be called with
+   * a new value. This is intended to be used as a getter/setter.
+   *
+   * The property has the same Epoch() function as a ReadOnlyProperty - the
+   * epoch value will be increased every time the property value is set.
+   *
+   * The property also has an Update() function, which can be called to manually
+   * increase the value returned by Epoch(). This is useful for mutable data
+   * types like arrays, which may be modified without explicitly calling the
+   * setter.
+   */
+  export interface Property<T> extends ReadOnlyProperty<T> {
+    (arg: T): T;
+    // Update increments the epoch manually; useful for certain data types like
+    // arrays, which may be modified without setting the property.
+    Update(): void;
+  }
+
+  /**
+   * Prop creates a new property with the given initial value.
+   */
+  export function Prop<T>(initial: T): Property<T> {
+    // obj stores the value of the property.
+    let obj: T = initial;
+
+    // epoch is a simple number which is incremented whenever the prop is set or
+    // when update() is called.
+    let epoch: number = 0;
+
+    let propFn: any = function(value?: T): T {
```

Comment:
It's the only way to construct a function with methods.  This is a very common javascript-ism, and typescript interfaces have explicit support for it, but there is currently no type-safe way to construct such an object in typescript.

Such is life.